Fix function require_splitter

dpino opened the pull request:
The function returns whether a V4V6 splitter app is required.  Before it checked whether an parameter `on-a-stick` was a command line argument, but this should not be the case as an on-a-stick setup is also possible when a configuration file has only one instance and one queue.

This caused that the function returned `true` if passing an `on-a-stick` argument to the command line, but returned `false` if running a lwaftr using only a configuration file.

After this fix I'll check whether the code path that not uses a splitter is correct (using vlan tags but different vlan tags), but that's another issue IMHO.

dpino committed:
Fix function require_splitter

The function returns whether a V4V6 splitter app is required.  Before it
checked whether an parameter `on-a-stick` was a command line argument,
but this should not be the case as an on-a-stick setup is also possible
when a configuration file has only one instance and one queue.

wingo commented on src/program/lwaftr/run/run.lua:
```diff
@@ -130,13 +130,10 @@ end
-      return internal_interface.vlan_tag == external_interface.vlan_tag
-   end
-   return false
+   local queue = select(3, lwutil.parse_instance(conf))
```
OMG I had no idea that "select" existed.  Weird!

I think you're probably right here, but I don't understand the whole picture; is it possible to configure two MAC addresses on one interface and use VMDq to present them as separate interfaces?  I think so, right?  In that case if the vlan tags are the same, no splitter needed.

Likewise we should allow queue.external_interface.device to be set, right?  It just *defaults* to be the same as the internal interface.  But we should allow users to specify it as being the same as the external interface.  In that case we should also go through the bump-in-the-wire configuration.  Right?

wingo replied:
lol i said "right?" many times.  that means i have little confidence in what i am saying :)

dpino replied:
The scenarios are the following when running on-a-stick:

- Both external and internal interfaces have the same VLAN tag value => a V4V6 splitter is required to separate incoming traffic to their corresponding apps.
- External and internal interfaces use different VLAN tags. Since we rely on the NIC to add and strip VLAN tag values, we need to virtualize the NIC so we can use different VLAN tag values per virtual NIC. Since we are already creating two virtual NICs (inet-side and b4-side), we don't need the splitter. 

I don't understand the `queue.external_interface.device` question. AFAIK, it's possible to set this value in a configuration. If an user doesn't set this value in a configuration, then it's nil. The value is only set if --v6 is passed as an argument.

In the former code, a splitter was not needed if we were in a non on-a-stick mode (that's why the argument was checked), but now we assume this function is only called if we're going to run on-a-stick (we know we're in on-a-stick mode because `queue.external_interface.device` is nil).

I agree it sounds a bit confusing, but I think it makes sense (I'm not 100% sure about `queue.external_interface.device`, maybe there's something more to do there).

wingo replied:
The `queue.external_interface.device` value can also be set if it's set in the configuration file

dpino replied:
OK, I think I understand now. I assumed that if `queue.external_interface.device` was set we are in a `bump-in-the-wire` scenario (non on-a-stick), but it could be the case that an user sets this value to the same value as the internal interface, which will be an `on-a-stick` scenario. I will check what happens in all these cases.

dpino committed:
Determine whether a configuration should run on-a-stick

A configuration should run on-a-stick always is on-a-stick command line
argument is uses, regardless of device setup in configuration.  The
arguments can only be used if the configuration has one instance.

In case on-a-stick argument is not used, a lwAFTR runs on-a-stick if
internal and external interfaces use the same device.

In case of running on-a-stick:
- Always use a V4V6 splitter app to separater IPv4 and IPv6 traffic.
- Except if internal and external interfaces use VLAN tags and the VLAN
  tags are different.

dpino commented:
I have pushed the following changes:

- If passing an `--on-a-stick` command line argument, the function that migrates configuration has to overwrite the value of `queue.external_interface.device` to the value of the `v4` device. Use case is a configuration that has a device for an instance and a different one for an external_interface.
- Added a function that determines whether to run on-a-stick if:
   - The configuration file has an instance device but not external_interface device.
   - Both interfaces are set and have the same value.
- If a configuration runs on-a-stick and each interface uses different VLAN tags, we skip the V4V6 splitter and setup two virtual interfaces for each NIC. However in this scenario, it's necessary that both interfaces have a different MAC address.

wingo commented on src/program/lwaftr/run/run.lua:
```diff
@@ -42,6 +42,12 @@ local function migrate_device_on_config(config, v4, v6)
          queue.external_interface.device = v6
       end
    end
+
```
WDYT about doing this at line 150 or so:

```lua
if opts['on-a-stick'] then assert(not v6); v6 = v4 end
```

then rely on the existing code doing the right thing.  WDYT?

wingo commented on src/program/lwaftr/setup.lua:
```diff
@@ -324,6 +324,9 @@ function load_on_a_stick(c, conf, args)
       link_source(c, v4v6..'.v4', v4v6..'.v6')
       link_sink(c, v4v6..'.v4', v4v6..'.v6')
    else
+      assert(queue.external_interface.mac ~= queue.internal_interface.mac,
```
Why is this requirement here?  Does it not work to have the same MAC address but different VLAN tags?

dpino replied:
Apparently it doesn't work. I saw 100% packets dropped with that configuration. Then it occurred to me to change the MAC addresses (and the data samples) and packets get routed.

I researched a bit to know the actual reason and I found something. Here's an excerpt from the Intel VMDq white paper that acknowledges this limitation:

_VMDq has certain limitations on sorting packets based on VLAN tags.  VMDq only sorts packets based on MAC address or VLAN tags. VMDq can not differentiate between two packets with the same MAC address and different VLAN tags._

https://www.intel.com/content/dam/www/public/us/en/documents/white-papers/vmdq-technology-paper.pdf

wingo replied:
Good find and great assert, then!

dpino committed:
Add on-a-stick assert and prepare migration

wingo reviewed (approved):
nice work!
